### ISSUE TITLE
dict key infered as object depending on param signature or condition order

### ISSUE THREAD

**Issue by pawamoy** [astral-sh/ty#2748]
### Summary

Sorry for the title, I really have no better idea on how to word this 🤔 Hopefully the reproduction will give you a better understanding. If this was already reported, I apologize, I failed to pick up the correct search terms 🙇 

```python
from typing import Any

from markdown import Extension

def _split_configs(
    markdown_extensions: list[str | dict[str, dict[str, Any]] | Extension],
) -> tuple[list[str | Extension], dict[str, dict[str, Any]]]:
    # Split markdown extensions and their configs from mkdocs.yml
    mdx: list[str | Extension] = []
    mdx_config: dict[str, dict[str, Any]] = {}
    for item in markdown_extensions:
        if isinstance(item, str):
            mdx.append(item)
        elif isinstance(item, dict):
            for key, value in item.items():
                mdx.append(key)
                mdx_config[key] = value
                break  # Only one item per dict
        elif isinstance(item, Extension):
            mdx.append(item)
    return mdx, mdx_config
```

```console
% uvx --with markdown ty check script.py                
error[invalid-argument-type]: Argument to bound method `append` is incorrect
  --> script.py:16:28
   |
14 |         elif isinstance(item, dict):
15 |             for key, value in item.items():
16 |                 mdx.append(key)
   |                            ^^^ Expected `str | Extension`, found `object`
17 |                 mdx_config[key] = value
18 |                 break  # Only one item per dict
   |
info: Method defined here
    --> stdlib/builtins.pyi:2843:9
     |
2841 |         """Return a shallow copy of the list."""
2842 |
2843 |     def append(self, object: _T, /) -> None:
     |         ^^^^^^       ---------- Parameter declared here
2844 |         """Append object to the end of the list."""
     |
info: rule `invalid-argument-type` is enabled by default

error[invalid-assignment]: Invalid subscript assignment with key of type `object` and value of type `object` on object of type `dict[str, dict[str, Any]]`
  --> script.py:17:17
   |
15 |             for key, value in item.items():
16 |                 mdx.append(key)
17 |                 mdx_config[key] = value
   |                 ^^^^^^^^^^^---^^^^-----
   |                            |      |
   |                            |      Expected value of type `dict[str, Any]`, got `object`
   |                            Expected key of type `str`, got `object`
18 |                 break  # Only one item per dict
19 |         elif isinstance(item, Extension):
   |
info: rule `invalid-assignment` is enabled by default

Found 2 diagnostics
```

Now if you just remove `| Extension` from the union of `markdown_extensions` in the signature, ty can infer the type of the dict key as string again (I guess):

```python
from typing import Any

from markdown import Extension

def _split_configs(
    markdown_extensions: list[str | dict[str, dict[str, Any]]],
) -> tuple[list[str | Extension], dict[str, dict[str, Any]]]:
    # Split markdown extensions and their configs from mkdocs.yml
    mdx: list[str | Extension] = []
    mdx_config: dict[str, dict[str, Any]] = {}
    for item in markdown_extensions:
        if isinstance(item, str):
            mdx.append(item)
        elif isinstance(item, dict):
            for key, value in item.items():
                mdx.append(key)
                mdx_config[key] = value
                break  # Only one item per dict
        elif isinstance(item, Extension):
            mdx.append(item)
    return mdx, mdx_config
```

```console
% uvx --with markdown ty check script.py
All checks passed!
```

Changing the conditions order also fixes the issue:

```python
from typing import Any

from markdown import Extension

def _split_configs(
    markdown_extensions: list[str | dict[str, dict[str, Any]] | Extension],
) -> tuple[list[str | Extension], dict[str, dict[str, Any]]]:
    # Split markdown extensions and their configs from mkdocs.yml
    mdx: list[str | Extension] = []
    mdx_config: dict[str, dict[str, Any]] = {}
    for item in markdown_extensions:
        if isinstance(item, str):
            mdx.append(item)
        elif isinstance(item, Extension):
            mdx.append(item)
        elif isinstance(item, dict):
            for key, value in item.items():
                mdx.append(key)
                mdx_config[key] = value
                break  # Only one item per dict
    return mdx, mdx_config
```

```console
% uvx --with markdown ty check script.py
All checks passed!
```

Is this expected 😊? The `Extension` class doesn't inherit from `dict` and doesn't provide an `items()` method (just in case).

### Version

ty 0.0.15

**Comment by AlexWaygood** [comment]
Thanks for the report, and sorry for the confusing error message! The cause here is described in #1578, and we have https://github.com/astral-sh/ty/issues/2221 to track improving our error message here
